substrate/primitives/runtime/src/generic/unchecked_extrinsic.rs

```diff
@@ -33,16 +33,206 @@ use codec::{Compact, Decode, Encode, EncodeLike, Error, Input};
+				let signature = Signature::decode(input)?;
+				let ext = Extension::decode(input)?;
+				Self::Signed(address, signature, 0, ext, version)
+			},
+			(EXTRINSIC_FORMAT_VERSION, SIGNED_EXTRINSIC) => {
+				let address = Address::decode(input)?;
+				let signature = Signature::decode(input)?;
+				let ext_version = ExtensionVersion::decode(input)?;
+				let ext = Extension::decode(input)?;
+				Self::Signed(address, signature, ext_version, ext, version)
+			},
+			(EXTRINSIC_FORMAT_VERSION, GENERAL_EXTRINSIC) => {
+				let ext_version = ExtensionVersion::decode(input)?;
+				let ext = Extension::decode(input)?;
+				Self::General(ext_version, ext)
+			},
+			(_, _) => return Err("Invalid transaction version".into()),
+		};
+
+		Ok(preamble)
+	}
+}
+
+impl<Address, Signature, Extension> Encode for Preamble<Address, Signature, Extension>
+where
+	Address: Encode,
+	Signature: Encode,
+	Extension: Encode,
+{
+	fn size_hint(&self) -> usize {
+		match &self {
+			Preamble::Bare(_) => EXTRINSIC_FORMAT_VERSION.size_hint(),
+			Preamble::Signed(address, signature, ext_version, ext, EXTRINSIC_FORMAT_VERSION) =>
+				EXTRINSIC_FORMAT_VERSION
+					.size_hint()
+					.saturating_add(address.size_hint())
+					.saturating_add(signature.size_hint())
+					.saturating_add(ext_version.size_hint())
+					.saturating_add(ext.size_hint()),
+			Preamble::Signed(
+				address,
+				signature,
+				_,
+				ext,
+				LOWEST_SUPPORTED_EXTRINSIC_FORMAT_VERSION,
+			) => LOWEST_SUPPORTED_EXTRINSIC_FORMAT_VERSION
+				.size_hint()
+				.saturating_add(address.size_hint())
+				.saturating_add(signature.size_hint())
+				.saturating_add(ext.size_hint()),
+			Preamble::General(ext_version, ext) => EXTRINSIC_FORMAT_VERSION
+				.size_hint()
+				.saturating_add(ext_version.size_hint())
+				.saturating_add(ext.size_hint()),
+			_ => {
+				// unreachable, versions are checked in the constructor
+				0
+			},
+		}
+	}
+
+	fn encode_to<T: codec::Output + ?Sized>(&self, dest: &mut T) {
+		match &self {
+			Preamble::Bare(extrinsic_version) => {
+				(extrinsic_version | BARE_EXTRINSIC).encode_to(dest);
+			},
+			Preamble::Signed(address, signature, ext_version, ext, EXTRINSIC_FORMAT_VERSION) => {
+				(EXTRINSIC_FORMAT_VERSION | SIGNED_EXTRINSIC).encode_to(dest);
+				address.encode_to(dest);
+				signature.encode_to(dest);
+				ext_version.encode_to(dest);
+				ext.encode_to(dest);
+			},
+			Preamble::Signed(
+				address,
+				signature,
+				_,
+				ext,
+				LOWEST_SUPPORTED_EXTRINSIC_FORMAT_VERSION,
+			) => {
+				(LOWEST_SUPPORTED_EXTRINSIC_FORMAT_VERSION | SIGNED_EXTRINSIC).encode_to(dest);
+				address.encode_to(dest);
+				signature.encode_to(dest);
+				ext.encode_to(dest);
+			},
+			Preamble::General(ext_version, ext) => {
+				(EXTRINSIC_FORMAT_VERSION | GENERAL_EXTRINSIC).encode_to(dest);
+				ext_version.encode_to(dest);
+				ext.encode_to(dest);
+			},
+			_ => {
+				// unreachable, versions are checked in the constructor
```

Comment:
At least add a `debug_assert!` that this is never hit.

Reply:
I added an error log instead, that should be visible all the time while not being dangerous in production.

Reply:
Logs are disabled in the production runtimes.